tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/temporal/TableauEncoder.scala

```diff
@@ -0,0 +1,423 @@
+                val auxVarExPrime = builder.prime(auxVarEx)
+
+                /*
+                  initialize __temporal_curNode
+
+                  /\ __temporal_curNode \in BOOLEAN
+                  /\ __temporal_curNode_globally = TRUE or /\ __temporal_curNode_finally = FALSE
+                 */
+                val initialValue = oper match {
+                  case TlaTempOper.box     => true
+                  case TlaTempOper.diamond => false
+                }
+                val auxVarInitEx = builder.eql(auxVarEx, builder.bool(initialValue))
+
+                /* update __temporal_curNode:
+
+                  /\ __temporal_curNode' \in BOOLEAN
+                  /\ __temporal_curNode_globally' \in BOOLEAN or __temporal_curNode_finally' \in BOOLEAN
+
+                                           outerOp                                          outerOp
+                  /\ __temporal_curNode <=> A /\ __temporal_curNode' or __temporal_curNode <=> A \/ __temporal_curNode'
+
+                                                    outerOp
+                  /\ __temporal_curNode_globally' <=> /\ __temporal_curNode_globally
+                                                      /\ (~InLoop' \/ A')
+                                                    outerOp      innerOp
+                  or
+                                                   outerOp
+                  /\ __temporal_curNode_finally' <=>  \/ __temporal_curNode_finally
+                                                      \/  (InLoop' /\ A')
+                                                    outerOp      innerOp
+                 */
+
+                /*
+                box and diamond differ in the inner and outer operators,
+                and whether the InLoop variable is negated in the conjunction
+                 */
+                val (outerOpTmp, innerOpTmp, innerInLoopEx) = oper match {
+                  case TlaTempOper.box     => (builder.and _, builder.or _, builder.not(loopEnc.inLoopPrime))
+                  case TlaTempOper.diamond => (builder.or _, builder.and _, loopEnc.inLoopPrime)
+                }
+
+                /*
+                outerOpTmp and innerOpTmp expect Seq[TBuilderInstruction] instead of TBuilderInstruction*,
+                so define new operators here to take TBuilderInstruction*
+                 */
+                def outerOp(args: TBuilderInstruction*): TBuilderInstruction = outerOpTmp(args)
+                def innerOp(args: TBuilderInstruction*): TBuilderInstruction = innerOpTmp(args)
+
+                /* __temporal_curNode_unroll' \in BOOLEAN */
+                val auxVarUpdateAssignmentEx = inBoolSet(builder.prime(auxVarEx))
+
+                /* __temporal_curNode <=> A /\ __temporal_curNode' */
+                val nodeVarUpdateConditionEx = builder.equiv(
+                    nodeVarEx,
+                    outerOp(
+                        argExs(0),
```

Comment:
You should make a note as to why args is nonempty and this won't throw. Also, possibly use `head`.